Add missing mini-expansion key item trading logic to Squintrox Dryeyes

MowFord opened the pull request:
**_I affirm:_**
- [x] I understand that if I do not agree to the following points by completing the checkboxes my PR will be ignored.
- [x] I understand I should leave resolving conversations to the LandSandBoat team so that reviewers won't miss what was said.
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/LandSandBoat/server/blob/base/CODE_OF_CONDUCT.md).
- [x] I have _**tested my code and the things my code has changed**_ since the last commit in the PR and will test after any later commits.

## What does this pull request do?

This is the partner to my Tenshodo Coffer PR. Squintrox logic is not fun to look at in this first commit, but I hope to utilize the setbit logic from tenshodo coffer and build a couple tables to shorten the core logic. 

It is tested and technically works, but I am not satisfied with the layout, just putting it here to get feedback on a path to the end result.

Edit: Testing is complete and I'm satisfied with the behavior

Notes about the logic for this NPC:
 - The order of page 1 `onTrigger` does NOT match the order of event update options and event finish options, so this is manually defined per option value. 
 - The order of page 2 `onTrigger` DOES match the order of event finish options, so this was simple to map `option - 299` to the `menuMetadata` index

## Steps to test these changes

This one is tricky to test as there's multiple layers of logic here:
- Trades:
  - Trading the required items for: `CRIMSON_KEY`, `WHITE_CORAL_KEY`, and `MOOGLE_KEY` are the simplest. 
  - Trade should only succeed if the `charVar` for that particular key item maps to a timestamp on the player's charvar less than now (this value is set to next jst midnight upon a successful trade)
- Page 1 key items:
  - Page 1 lists the key items available based on
    - The respective mini expansion being complete (mission log shows the `FIN` mission as active)
    - the corresponding timestamp in the player's `charVar` player variable.
    - player not currently having the key item
  - if the player chooses that key item, the requirements are checked:
    - if the player has all required key items (most require full list, `BIRD_KEY` only requires 3 of the 6 key items) they receive the key item and all req (and prereq) key items are removed
    - if the player doesn't meet the requirements, info is given:
      - for all but `BLACK_CORAL_KEY` and `MOOGLE_KEY`, a simple text is given stating the requirements for obtaining the repeat key item
        - `MOOGLE_KEY`: the particular enfeeblement kit required is chosen/listed
        - `BLACK_CORAL_KEY`: the particular zone for digging is chosen/listed
- Page 2 key items:
  - These are all prerequisite key items for completing the mini-expansion battles again to obtain the key item rewards
  - They are not displayed if any is true
    - The respective mini expansion is not complete (mission log shows the `FIN` mission as active)
    - you currently have _ALL_ the key items 
    - you're on cooldown due to the `charVar` timestamp on your player
  - If chosen, they cost a varied amount of gil/beastmen seals, with custom messages for each type of result

Various screenshots of testing

![image](https://github.com/LandSandBoat/server/assets/131182600/4d7f8420-0668-4aa4-8327-0fc69f1545e7)

![image](https://github.com/LandSandBoat/server/assets/131182600/b49f1a16-c54d-4e23-ae9c-ce4fed326d74)

![image](https://github.com/LandSandBoat/server/assets/131182600/118b0185-6cf2-4177-ba05-1d3fc4c9d57e)

MowFord committed:
Initial submission for squintrox

zach2good commented on scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua:
```diff
@@ -3,117 +3,534 @@
+    local finishedASA  = player:getCurrentMission(xi.mission.log_id.ASA) == xi.mission.id.asa.A_SHANTOTTO_ASCENSION_FIN
+
+    -- Reminder that "True" means the option should be excluded from the player's menu
+    local arg1 =
+        ((finishedACP                                              or
+            player:hasKeyItem(xi.ki.CRIMSON_KEY))                  and
+                2 or 0) +
+        ((finishedACP                                              or
+            player:hasKeyItem(xi.ki.VIRIDIAN_KEY))                 and
+                4 or 0) +
+        ((finishedAMK                                              or
+            player:hasKeyItem(xi.ki.WHITE_CORAL_KEY))              and
+                8 or 0) +
+        ((finishedAMK                                              or
+            player:hasKeyItem(xi.ki.BLUE_CORAL_KEY))               and
+                16 or 0) +
+        ((finishedAMK                                              or
+            player:hasKeyItem(xi.ki.BLACK_CORAL_KEY))              and
+                32 or 0) +
+        ((finishedASA                                              or
+            player:hasKeyItem(xi.ki.MOOGLE_KEY)                    and
+            now >= player:getCharVar('LastMoogleKey'))             and
+                64 or 0) +
+        ((finishedASA                                              or
+            player:hasKeyItem(xi.ki.BIRD_KEY)                      and
+            now >= player:getCharVar('LastBirdKey'))               and
+                128 or 0) +
+        ((finishedASA                                              or
+            player:hasKeyItem(xi.ki.BOMB_KEY)                      and
+            now >= player:getCharVar('LastBombKey'))               and
+                256 or 0)
```
This looks suspiciously like it could be broken out into a table and then iterated over

```lua
-- { expansion, keyItem, maskValue }
local lookup =
{
    { <mark ACP condition somehow>, xi.ki.CRIMSON_KEY, bit.lshift(1, 1) },
    ...
}

local arg1 = 0
for _, entry in pairs(lookup) do
    local hasCompletedExpansion = player:checkCompleteEtc...(entry[1])
    local hasRelevantKeyItem        = player:hasKI...(entry[2])
    if hasCompletedExpansion or hasRelevantKeyItem then
        local maskValue = entry[3]
        arg1 = arg1 + maskValue
    end
end
```

MowFord replied:
The real lpt is always in the comments

zach2good commented on scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua:
```diff
@@ -3,117 +3,534 @@
-        xi.settings.main.ENABLE_AMK == 0 and
-        xi.settings.main.ENABLE_ASA == 0
-    then
+    local now          = os.time()
+    local nextMidnight = getMidnight()
```
These two aren't used in here

MowFord replied:
Good call I copy pasted definitions for consistent naming

zach2good commented on scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua:
```diff
@@ -3,117 +3,534 @@
-            player:addKeyItem(xi.ki.CRIMSON_KEY)
-            player:setCharVar('LastCrimsonKey', getMidnight())
+            player:addKeyItem(ki)
+            player:setCharVar('LastCrimsonKey', nextMidnight)
```
With the different key trades, you can probably write some helper funcs to deal with the repeated logic

MowFord replied:
Fair point

zach2good commented on scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua:
```diff
@@ -3,117 +3,534 @@
+    local hasMoogleItems        = kit ~= 0 and
+                                    trade:hasItemQty(kit, 1) and
+                                    count == 1
+    local hasCrimsonItems       = trade:hasItemQty(xi.item.SEEDSPALL_LUX, 1)        and
+                                    trade:hasItemQty(xi.item.SEEDSPALL_LUNA, 1)     and
+                                    trade:hasItemQty(xi.item.SEEDSPALL_ASTRUM, 1)   and
+                                    count == 3
+    local hasWhiteCoralItems    = trade:hasItemQty(xi.item.ORCISH_PLATE_ARMOR, 1)   and
+                                    trade:hasItemQty(xi.item.QUADAV_BACKSCALE, 1)   and
+                                    trade:hasItemQty(xi.item.YAGUDO_CAULK, 1)       and
```
This giant block of things defined at the top is quite unsightly, but this is quite an overloaded NPC - so the option are to use these blocks inline if they're only used one (like here), or to break them out into local functions that takes the trade and/or the player as args and returns a bool

MowFord replied:
Figured it was better to define all the complex requirements in one place at the top, but I could put them directly in the if blocks if that's preferred

zach2good commented on scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua:
```diff
@@ -3,117 +3,534 @@
+            player:tradeComplete()
+            player:addKeyItem(ki)
+            player:setCharVar('LastMoogleKey', nextMidnight)
+            player:setCharVar('ASA_kit', 0)
```
I don't know the current state of ASA in LSB, but this will eventually be replaced by the numbered vars of the interaction framework, so I'd look into what those will look like for their eventual migration

MowFord replied:
I figured this would just be a placeholder until that happens (it'll likely be me and flibe doing them)

zach2good commented on scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua:
```diff
@@ -3,117 +3,534 @@
+        hasEnfeebKit and
+        finishedASA
+    then
+        local ki = xi.ki.MOOGLE_KEY
```
This extraction doesn't save much, just keep the usage inline

zach2good commented on scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua:
```diff
@@ -3,117 +3,534 @@
+-- validation of prereqs done here, event option updated based on checks to be used in onEventFinish
 entity.onEventUpdate = function(player, csid, option, npc)
+    if csid == 323 then
+        if option == 100 then -- Viridian Key
```
Each of the blocks in here are quite unique, and take up quite varying amounts of space - I'd probably break all of them out into their own functions

zach2good commented on scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua:
```diff
@@ -3,117 +3,534 @@
+            local finishedASA = player:getCurrentMission(xi.mission.log_id.ASA) == xi.mission.id.asa.A_SHANTOTTO_ASCENSION_FIN
+
+            -- Reminder that "True" means the option should be excluded from the player's menu
+            local arg1 = (not finishedACP                                                or
+                             (player:hasKeyItem(xi.ki.SEEDSPALL_ROSEUM)                 and
+                             player:hasKeyItem(xi.ki.SEEDSPALL_CAERULUM)                and
+                             player:hasKeyItem(xi.ki.SEEDSPALL_VIRIDIS))                and
+                                 2 or 0) +
+                         (not finishedACP                                                or
+                             (player:hasKeyItem(xi.ki.MARK_OF_SEED))                    and
+                                 4 or 0) +
+                         (not finishedACP                                                or
```
I'd look into the pattern I mentioned before about this being a data table that's then iterated over. Remember you can nest tables, so you can have a table of KI requirements as an entry in the main table

MowFord replied:
I think that is the route I will take. The same data is also used in the next event function in the same order so it should extend nicely

zach2good commented on scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua:
```diff
@@ -3,117 +3,534 @@
+            player:delSeals(15, 0)
+            player:delGil(1500)
+
+            for i, ki in pairs({
```
For things like this, I'd define the table first and then use it, rather than inline in the loop definition

MowFord replied:
I like it

zach2good commented on scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua:
```diff
@@ -3,117 +3,534 @@
+            player:delSeals(20, 0)
+            player:delGil(2000)
+
+            for i, ki in pairs({
```
For things like this, I'd define the table first and then use it, rather than inline in the loop definition.

Think about how these sorts of things are specific to the expansion they come from. Can they be added to the helpers files for those expansions and then referenced here like `xi.asaHelpers.keyItemsWhatever`?

MowFord replied:
Oh interesting idea. I think I'd rather keep this file self contained tbh though. Since except for a could char var checks it's mostly self contained

Would it be ok to define the single table in this file with all of it and then update later when we audit the expansions?

In doing the expansions on wings it was very frustrating not having this file fleshed out first

MowFord commented:
I have learned to run CI checks locally (go me), I also plan on investigating the commit ammending/force push to not require squashing every PR of mine. I don't mind if you squash this one or I can go back and combine the two commits myself.

I hope i didn't get too in my head with the `menuMetadata` table of key item info. I think I struck a balance between not repeating code and readability.

Note that I _have not_ checked the "i'm fully tested this code" box as I'm waiting on feedback on the formatting before I thoroughly check every iteration.

zach2good commented on scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua:
```diff
@@ -3,57 +3,348 @@
+    local ki     = xi.ki.MOOGLE_KEY
+    local asaKit = player:getCharVar(menuMetadata[1][ki].reqItemCharVar)
+
+    if -- Crimson Key: Trade Seedspall's Lux, Luna, Astrum
+        trade:hasItemQty(menuMetadata[1][xi.ki.CRIMSON_KEY].reqItems[1], 1)   and
+        trade:hasItemQty(menuMetadata[1][xi.ki.CRIMSON_KEY].reqItems[2], 1)   and
+        trade:hasItemQty(menuMetadata[1][xi.ki.CRIMSON_KEY].reqItems[3], 1)   and
+        count == 3                                                            and
+        player:getCurrentMission(xi.mission.log_id.ACP) == xi.mission.id.acp.A_CRYSTALLINE_PROPHECY_FIN
+    then
+        ki = xi.ki.CRIMSON_KEY
+        tradeForKeyItem(player, trade, ki)
+    elseif -- White Coral Key: orcish plate armor, quadav backscale, yagudo caulk
+        trade:hasItemQty(menuMetadata[1][xi.ki.WHITE_CORAL_KEY].reqItems[1], 1)   and
+        trade:hasItemQty(menuMetadata[1][xi.ki.WHITE_CORAL_KEY].reqItems[2], 1)   and
+        trade:hasItemQty(menuMetadata[1][xi.ki.WHITE_CORAL_KEY].reqItems[3], 1)   and
+        count == 3                                                                and
+        player:getCurrentMission(xi.mission.log_id.AMK) == xi.mission.id.amk.A_MOOGLE_KUPO_DETAT_FIN
+    then
+        ki = xi.ki.WHITE_CORAL_KEY
+        tradeForKeyItem(player, trade, ki)
+    elseif -- Moogle Key: trade proper enfeebling kit
+        asaKit ~= 0                                                    and
+        trade:hasItemQty(asaKit, 1)                                    and
+        count == 1                                                     and
+        player:getCurrentMission(xi.mission.log_id.ASA) == xi.mission.id.asa.A_SHANTOTTO_ASCENSION_FIN
+    then
```
The arbitrary gaps between the statement and the and/or keywords aren't needed here. Only put in extraneous whitespace if you're aligning things to something: like variable names or the = operators

MowFord replied:
Ok so just

```
Statement and
Longer statement and
Short and
Blah
```
?

zach2good replied:
Yessir

zach2good commented on scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua:
```diff
@@ -3,57 +3,348 @@
+local function verifyReqKeyItems(player, ki)
+    local entry = menuMetadata[1][ki]
+    local count = 0
+    for i, reqKeyItem in pairs(entry.reqKeyItems) do
```
If you're not using i, you should use _ to signal that the key isn't being used

zach2good commented on scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua:
```diff
@@ -3,57 +3,348 @@
+                xi.ki.CHUNK_OF_SMOKED_GOBLIN_GRUB,
+            }
+        },
+        [xi.ki.WHITE_CORAL_KEY] =
+        {
+            expansion = 2,
+            charVar   = 'LastWhiteCoralKey',
+            reqItems =
+            {
+                xi.item.ORCISH_PLATE_ARMOR,
+                xi.item.QUADAV_BACKSCALE,
+                xi.item.YAGUDO_CAULK,
+            },
+        },
+        [xi.ki.BLUE_CORAL_KEY] =
+        {
+            expansion = 2,
+            charVar   = 'LastBlueCoralKey',
+            reqKeyItems =
+            {
+                xi.ki.STURDY_METAL_STRIP,
+                xi.ki.PIECE_OF_RUGGED_TREE_BARK,
+                xi.ki.SAVORY_LAMB_ROAST,
+            },
+        },
+        [xi.ki.BLACK_CORAL_KEY] =
+        {
+            expansion = 2,
+            charVar   = 'LastBlackCoralKey',
+            reqKeyItems =
+            {
+                xi.ki.MOLDY_WORMEATEN_CHEST,
+            },
+        },
+        [xi.ki.MOOGLE_KEY] =
+        {
+            expansion      = 3,
+            charVar        = 'LastMoogleKey',
+            reqItemCharVar = 'ASA_kit',
+            reqItems =
```
This table is nice, but the expansion, charVar and reqItemCharVar keys aren't correctly aligned like the rest of the table - choose one style and stick to it

zach2good reviewed (commented):
Liking the new layout 👍

zach2good commented on scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua:
```diff
@@ -65,53 +367,162 @@ entity.onTrigger = function(player, npc)
+                player:delSeals(entry.costSeals, 0)
+                player:delGil(entry.costGil)
+
+                for i, ki in pairs(entry.relevantKeyItems) do
```
i -> _

MowFord replied:
:( sorry

MowFord committed:
Squintrox Dryeyes functionality:

table to define mappings of key items to requirements
generalized functions where code was repeated

claywar commented on scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua:
```diff
@@ -3,58 +3,360 @@
+    local count = 0
+    for _, reqKeyItem in pairs(entry.reqKeyItems) do
+        if reqItemCount > 0 then
+            print(reqKeyItem)
```
Debug print can be removed

claywar commented on scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua:
```diff
@@ -65,53 +367,162 @@ entity.onTrigger = function(player, npc)
+
+                for _, ki in pairs(entry.relevantKeyItems) do
+                    if not player:hasKeyItem(ki) then
+                        player:addKeyItem(ki)
```
These two lines can be replaced by the npcUtil function (which can handle either single input or tables of KIs)

MowFord replied:
Good catch.

claywar commented on scripts/zones/Port_Jeuno/npcs/Squintrox_Dryeyes.lua:
```diff
@@ -65,53 +367,162 @@ entity.onTrigger = function(player, npc)
+            takeReqKeyItems(player, xi.ki.BIRD_KEY)
+        elseif option >= 300 and option <= 308 then
+            local entry = menuMetadata[2][option - 299]
+            if player:getSeals(0) < entry.costSeals then
```
Not as familiar with this NPC, but does he really pull from stored seals and don't require a trade?

MowFord replied:
Sure does!

https://ffxiclopedia.fandom.com/wiki/Talk:Squintrox_Dryeyes

> She doesn't take the beastmen seals from you, she takes them from [Shami](https://ffxiclopedia.fandom.com/wiki/Shami). That's why she mentions that she and Shami have a deal. Trade the seals to him and you'll be able to buy the key items from Squintrox. --[Kyrie](https://ffxiclopedia.fandom.com/wiki/User:Kyrial) 02:28, June 30, 2010 (UTC)

zach2good replied:
And still does:
<img width="506" alt="image" src="https://github.com/LandSandBoat/server/assets/1389729/46b13a0f-ac04-4e78-ab60-77898e3e99be">

MowFord committed:
Simplify logic using the npcUtil.giveKI func

MowFord commented:
I've updated the things clay found, but decided not to force push since zach already approved it

zach2good commented:
Don't worry about force pushing things etc. it doesn't dismiss the approval (I think) and GitHub is pretty good at showing `New changes since you last viewed` and even keeps copies of earlier commits.

We're also pretty good at not yeeting merges in if other people have ongoing conversations

MowFord commented:
Sorry to hijack a different PR but I just wanted to pose a quick question about the AMK work we'll be starting up.

Once this PR is merged I'd like to update the feature branch with `base` (to capture this and tenshodo coffer updates). Would you like me to `git merge base` for my first PR or would you like to update the feature branch before I submit the first PR? I'm not well versed in the best ways to handle merge commits.

zach2good commented:
Once this merges I'll rebase the feature branch, so you can start targetting that without having to do any merges 👍
